### ISSUE TITLE
Docking specific example

### ISSUE THREAD

**Issue by axd2201** [ocornut/imgui#2304]
**Version/Branch of Dear ImGui:**

Version: Latest
Branch: Docking

**Back-end/Renderer/Compiler/OS**

Back-ends: Directx9 Renderer
Operating System: Windows 10

**My Issue/Question:**

Short version: Does not remenber the selected tab when stopped being rendered and starts again.

I have multiple tabs inside a single window, let's call them A B and C, if i open menu A is selected, i select B and close, when i open it again it opens with A as being selected/open instead of B

Well this is kinda of an extra question, that i searched so much and couldnt find anywhere, is the any way to create many Windows, and dock/tab them to a single window programatically?

**Comment by ocornut** [comment]
> I have multiple tabs inside a single window, let's call them A B and C, if i open menu A is selected, i select B and close, when i open it again it opens with A as being selected/open instead of B

This is ambiguous, pleas provide a repro.

**Comment by axd2201** [comment]
@ocornut Hello, sorry for the bad explanation and lack of Demo and Source Code, i was on short time when reporting this, i hope with the extra info that i added to the issue it self will help you figure out the issue, will try to do a better job in the future, if you have any issues or want my help just tell me will be happy to glad, thanks for your time have a nice day.

**Comment by ocornut** [comment]
> sorry for the bad explanation and lack of Demo and Source Code, i was on short time when reporting this

This is not an acceptable excuse. I'm providing a service to the community and there are 2304 issues on this tracker. The least you can do when asking for help is to provide me with the information I am requesting.

There are various possible ambiguity - I can't tell from your description if you are you using explicit tab bars, or windows that you docked/merged together. How do you hide and restore them etc? All those questions have an effect on the code flow. The code and gif allows to lift this ambiguity.

So - there is indeeed a bug with the current code which I can repro using:

```
static bool toggle = false;
ImGui::Checkbox("toggle", &toggle);
if (toggle)
{
    ImGui::Begin("AAAA");
    ImGui::Text("Hello from A window!");
    if (ImGui::Button("Close Me")) {}
    ImGui::End();

    ImGui::Begin("BBBB");
    ImGui::Text("Hello from B window!");
    if (ImGui::Button("Close Me")) {}
    ImGui::End();

    ImGui::Begin("CCCC");
    ImGui::Text("Hello from C window!");
    if (ImGui::Button("Close Me")) {}
    ImGui::End();
}
```

Variation of this bug have been coming up regularly. I'm looking at this and I'll try to add dedicated tests in my testing system when I find a solution for it.

> Well this is kinda of an extra question, that i searched so much and couldnt find anywhere, is the any way to create many Windows, and dock/tab them to a single window programatically?

This is discussed in the docking thread. It's not trivially faisible at the moment, but there's a WIP internal DockBuilder API which is not easy to use and going to evolve.

**Comment by ocornut** [comment]
FYI this issue isn't actually what I thought it was - it's not related to Docking in particular, but simply to the fact that when multiple windows are re-opened on the same frame they all reclaim the focus in order of submission.

So if you do e.g.

Frame N+0
- submit A (takes focus)
- submit B (takes focus)
- submit C (takes focus)

Frame N+100
- focus B

Frame N+200
- stop submitting them for a brief moment 

Frame N+300
- submit A (takes focus)
- submit B (takes focus)
- submit C (takes focus)

Then at the end C is focused again.
This is regardless of any docking consideration. In this situation the Docking tab bar just follows the "currently focused" window.

I'd like to eventually enable for remembering the respeective z-order and focus-order of windows so multiple windows re-enabled in the same frame can restore their relative order. This is not a trivial change so I don't know when I'll be able to do it.

**Comment by nem0** [comment]
I have a simple hack that kind of fixes this for docking - specifically the part where active tab is restored from imgui.ini. This is not in any way a correct fix for the general focus issue. Maybe somebody will find this useful. 

[restore_dock.zip](https://github.com/ocornut/imgui/files/5932278/restore_dock.zip) it's a patch file made on 22d9a61b3349be48844dd17a2da619c5ea6c05eb

**Comment by ocornut** [comment]
I looked at this and found two bizarre things:

- Can't get your patch to work in my test case (which is submitting 4 windows, then I dock them, focus one, shutdown/reopen app). I'm sure it does work in your situation but it shows how there's quite some complexity to that system.
- Very bizarrely I realize we DO have code saving this info in .ini file and attempting to restoring it. I honestly forgot this code even existed, it's seeing your patch that reminded me that data was saved. I don't think the restoring worked (or in only some cases) for a while.

I'm going to have to do some digging, design a battery of tests and see if we can fix this properly for docking or even for regular windows.

**Comment by nem0** [comment]
After I posted the patch file, I found some issues with it. I'm running slightly changed version https://github.com/nem0/LumixEngine/blob/master/external/imgui/imgui.cpp#L13452 for 20 days, no issues with it so far.  New version does not use `LastFrameActive` to detect if `DockNodeUpdateTabBar` is being called for the first time.

**Comment by ocornut** [comment]
Just to confirm that we haven't forgot about that.

nem0: I was hesitant to pull your patch as-is because the "first time" thing leads to further inconsistencies when hiding and reenabling dockspaces multiple times during the session, and we have a few users who do this regularly. Right now I'm waiting on a new series of regression tests to be written related to focus and tab sync as there is a bunch of code smell in there, and we then will tackle and clean all of this. Thanks for your patience.

**Comment by nem0** [comment]
No worries. I did not submit PR because I expected it not to be final/correct solution. I still have focus issues for "not the first time" cases, e.g. after minimizing and restoring window. They are not as annoying though and I'm pretty happy with current state.

**Comment by ocornut** [comment]
_Duke Nukem Forever release update_: I was about to push a fix for this last month then stumbled on a severe flaw/side-effect and had to put it on hold. Been on and off continuously following the thread of making the right fix for that. In fact, the variety of seemingly random Navigation fixes in 1.83 are direct side-effects of following that thread (basically the non-problematic fix involve processing focus of reappearing windows at end of frame and there are a few stuff left which have remaining dependencies in `Begin()`).

**Comment by nem0** [comment]
Since my old hack does not work anymore, here's a new one in case someone finds this useful
[dock_restore_active.patch](https://github.com/ocornut/imgui/files/12139701/dock_restore_active.patch) vs f8f805f04631323c8a75accc009eb9701f5ca027

**Comment by ocornut** [comment]
I have pushed 61acb34e which afaik should lead to the same result as your patch nem0 afaik non-focused node should correctly restore selected tabs. (focused nodes still wont).

The bulk of this topic is still open but this is a good quality of life improvement in the meanwhile.

And reworked a test bed for that https://github.com/ocornut/imgui_test_engine/commit/e4ea74f197fe1ddf62b9c8cbe45570666c3c10da

**Comment by nem0** [comment]
I tested 61acb34e32c3710c4c5a56f5e41b999742e6780e, it seems to help but does not completely fix my issue (neither does my patch). Some nodes are restored correctly and some are not.

**Comment by ocornut** [comment]
The ones that are restored correctly are the one which are the time of display do NOT host the focused window.
